internal/sync2/cycle.go

```diff
@@ -124,9 +143,10 @@ func (cycle *Cycle) Run(ctx context.Context, fn func(ctx context.Context) error)
 
 // Close closes all resources associated with it.
 func (cycle *Cycle) Close() {
-	cycle.Stop()
-	<-cycle.stop
-	close(cycle.control)
+	if cycle.runningState.IsRunning() {
```

Comment:
Could multiple calls to IsRunning() here race?  Should that logic be moved into here to hold the Lock() longer?

Reply:
We do need a mutex as the tests fail otherwise with a data race. Now for the Lock, I don't really know.

Reply:
IsRunning returns a snapshot of state not protected by the mutex.  Cycle.Stop() will run twice if Close() is called from two different threads

Reply:
Close shouldn't be called concurrently in the first place.